service/internal/status/status.go

```diff
@@ -0,0 +1,167 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package status // import "go.opentelemetry.io/collector/service/internal/status"
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"go.opentelemetry.io/collector/component"
+)
+
+// onTransitionFunc receives a component.StatusEvent on a successful state transition
+type onTransitionFunc func(*component.StatusEvent)
+
+// errInvalidStateTransition is returned for invalid state transitions
+var errInvalidStateTransition = errors.New("invalid state transition")
+
+// fsm is a finite state machine that models transitions for component status
+type fsm struct {
+	current      *component.StatusEvent
+	transitions  map[component.Status]map[component.Status]struct{}
+	onTransition onTransitionFunc
+}
+
+// transition will attempt to execute a state transition. If it's successful, it calls the
+// onTransitionFunc with a StatusEvent representing the new state. Returns an error if the arguments
+// result in an invalid status, or if the state transition is not valid.
+func (m *fsm) transition(status component.Status, options ...component.StatusEventOption) error {
+	if _, ok := m.transitions[m.current.Status()][status]; !ok {
+		return fmt.Errorf(
+			"cannot transition from %s to %s: %w",
+			m.current.Status(),
+			status,
+			errInvalidStateTransition,
+		)
+	}
+
+	ev, err := component.NewStatusEvent(status, options...)
+	if err != nil {
+		return err
+	}
+
+	m.current = ev
+	m.onTransition(ev)
+
+	return nil
+}
+
+// newFSM creates a state machine with all valid transitions for component.Status.
+// The initial state is set to component.StatusNone.
+func newFSM(onTransition onTransitionFunc) *fsm {
+	initial, _ := component.NewStatusEvent(component.StatusNone)
+	return &fsm{
+		current:      initial,
+		onTransition: onTransition,
+		transitions: map[component.Status]map[component.Status]struct{}{
+			component.StatusNone: {
+				component.StatusStarting: {},
+			},
+			component.StatusStarting: {
+				component.StatusOK:               {},
+				component.StatusRecoverableError: {},
+				component.StatusPermanentError:   {},
+				component.StatusFatalError:       {},
+				component.StatusStopping:         {},
+				component.StatusStopped:          {},
```

Comment:
I tend to think we shouldn't be able to go directly to Stopped, without going through Stopping.

Reply:
It was an oversight, and cascaded via copypasta. Thanks for catching this.